### ISSUE TITLE
sngrep - 1.4.8 crashed in Debian bullseye

### ISSUE THREAD

**Issue by lexxua** [irontec/sngrep#390]
Hi Sngrep team,
I noticed that sngrep starts crashing in Debian bullseye. 
I have following version installed from Debian repo:
sngrep -V
sngrep - 1.4.8
Copyright (C) 2013-2018 Irontec S.L.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
 * Compiled with GnuTLS support.
 * Compiled with Wide-character support.
 * Compiled with Perl Compatible regular expressions support.
 * Compiled with IPv6 support.
 * Compiled with EEP/HEP support.

Written by Ivan Alonso [aka Kaian]

apt-cache policy libssl1.1
libssl1.1:
  Installed: 1.1.1k-1+deb11u1
  Candidate: 1.1.1k-1+deb11u1

```
Here is my bt full:
`GNU gdb (Debian 10.1-1.7) 10.1.90.20210103-git
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/sngrep...
Reading symbols from /usr/lib/debug/.build-id/0f/2def574add6d175065427b6594139bf729e9bf.debug...

warning: Can't open file socket:[1989568] which was expanded to socket:[1989568] during file-backed mapping note processing
[New LWP 856346]
[New LWP 856345]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `sngrep -c'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000556acecb0c7e in memcpy (__len=32, __src=<optimized out>, __dest=0x7f1098028b68) at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:34
34	/usr/include/x86_64-linux-gnu/bits/string_fortified.h: No such file or directory.
[Current thread is 1 (Thread 0x7f109ed13700 (LWP 856346))]
(gdb) bt full
#0  0x0000556acecb0c7e in memcpy (__len=32, __src=<optimized out>, __dest=0x7f1098028b68) at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:34
No locals.
#1  tls_process_record_ssl2 (conn=0x7f1098028b30, payload=payload@entry=0x7f109802cc50 "\200", outl=0x7f109ed0d8bc, out=0x7f109ed0d8c8, len=396) at capture_gnutls.c:489
        clienthello = <optimized out>
        random = <optimized out>
        record_len_len = <optimized out>
        record_len = <optimized out>
        record_type = <optimized out>
        fragment = <optimized out>
        flen = <optimized out>
#2  0x0000556acecb23f0 in tls_process_record_ssl2 (outl=0x7f109ed0d8bc, out=0x7f109ed0d8c8, len=396, payload=0x7f109802cc50 "\200", conn=<optimized out>) at capture_gnutls.c:453
        record_len_len = <optimized out>
        record_len = <optimized out>
        record_type = <optimized out>
        fragment = <optimized out>
        flen = <optimized out>
        clienthello = <optimized out>
        random = <optimized out>
#3  tls_process_segment (packet=packet@entry=0x7f109800aba0, tcp=tcp@entry=0x7f109ed0d984) at capture_gnutls.c:389
        conn = <optimized out>
        payload = 0x7f109802cc50 "\200"
        size_payload = 396
        out = 0x7f109802cdf0 ""
        outl = 396
        ip_src = {s_addr = 2922651308}
        ip_dst = {s_addr = 2939428524}
        sport = 24007
        dport = 49107
        tlsserver = {ip = '\000' <repeats 45 times>, port = 0}
#4  0x0000556acecaf5a6 in parse_packet (info=0x556acf3377d0 "\001", header=<optimized out>, packet=<optimized out>) at capture.c:354
        capinfo = 0x556acf3377d0
        udp = <optimized out>
        udp_off = <optimized out>
        tcp = 0x7f109ed0d984
        tcp_off = <optimized out>
        data = "\000\004\000\001\000\006\332\060\245\206C`\000\000\b\000E\000\001\300\335}@\000@\006\232 \254\036\064\256\254\036\064\257]\307\277\323\356\037k7\253\363\253g\200\030\001\375\303L\000\000\001\001\b\n\262\234\363\241u\335ma\200\000\001\210\000\000%I\000\000\000\001", '\000' <repeats 35 times>, "\001\000\000\000\aGF-DUMP\000\000\000\000\000\a[\270m\000\000\000\000\000\000\000\001\000\000\000\001\000\000\000\021GlusterD svc peer\000\000\000\000\000\000\000\000\022\345\245\000\000\000\000\000\000\000\002\000\000\000\001\000\000\000\032Glus"...
        payload = 0x7f109ed0d9a4 "\200"
        size_capture = 464
        size_payload = 396
        pkt = 0x7f109800aba0
#5  0x00007f109f83ac05 in pcap_handle_packet_mmap (handle=handle@entry=0x556acf337820, callback=callback@entry=0x556acecaf430 <parse_packet>, user=user@entry=0x556acf3377d0 "\001", frame=frame@entry=0x7f109ed44230 "", tp_len=<optimized out>, tp_mac=<optimized out>, tp_snaplen=448, tp_sec=1644341579,
    tp_usec=439509, tp_vlan_tci_valid=0, tp_vlan_tci=0, tp_vlan_tpid=33024) at ./pcap-linux.c:3965
        handlep = 0x556acf337aa8
        bp = <optimized out>
        sll = <optimized out>
        pcaphdr = {ts = {tv_sec = 1644341579, tv_usec = 439509}, caplen = 464, len = 464}
        snaplen = <optimized out>
        utsname = {sysname = '\000' <repeats 64 times>, nodename = '\000' <repeats 64 times>, release = '\000' <repeats 64 times>, version = '\000' <repeats 64 times>,
          machine = '\000' <repeats 28 times>, "\002\000\000\000\000\000\000\000\023\324s\237\020\177\000\000\000\000\000\000\000\000\000\000\001\000\000\000\020\177\000\000\002\000\000\000",
          domainname = "\000\000\000\060\212\216\237\377\377\377\377f\000\000\000\000\000\000\000\364\202\203\237\020\177", '\000' <repeats 18 times>, "\004\000\000\000\001\000\001\000\003\000\000\000\001\000\000\000\000\000\000\000\000"}
#6  0x00007f109f83b074 in pcap_read_linux_mmap_v3 (handle=0x556acf337820, max_packets=<optimized out>, callback=0x556acecaf430 <parse_packet>, user=0x556acf3377d0 "\001") at ./pcap-linux.c:4115
        tp3_hdr = 0x7f109ed44230
        packets_to_read = 136
        handlep = 0x556acf337aa8
        h = {h2 = 0x7f109ed14000, h3 = 0x7f109ed14000, raw = 0x7f109ed14000 ""}
        pkts = 312
        ret = <optimized out>
        again = <optimized out>
#7  0x00007f109f841b0e in pcap_loop (p=0x556acf337820, cnt=cnt@entry=-1, callback=callback@entry=0x556acecaf430 <parse_packet>, user=user@entry=0x556acf3377d0 "\001") at ./pcap.c:2904
        n = <optimized out>
#8  0x0000556acecad8fc in capture_thread (info=0x556acf3377d0) at capture.c:883
        capinfo = 0x556acf3377d0
#9  0x00007f109f817ea7 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
No symbol table info available.
#10 0x00007f109f747def in clone () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
quit)`

Thank you for help!

**Comment by Kaian** [comment]
Hi,

Looks like the crash ocurs while decrypting a TLS packet.
The command line has no -k parameter so I assume it's configured in your sngreprc, right?

Can you provide a sample pcap and key that reproduces the issue?

Thanks!

**Comment by marcocapetta** [comment]
Hi @Kaian,

I'm continuing the description of the problem reported by @lexxua 
I can confirm the keyfile is configured in the configuration file using "set capture.keyfile"
If we remove the keyfile from config sngrep doesn't crash anymore.

Tested also manually compiling last available version 1.4.10 with GnuTLS support and it crashes as well.

We also tried to manually compile the last available version 1.4.10 with OpenSSL support and in this case we didn;t have any crash.

Thanks
